pkg/storage/edv/encryptedformatter.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package edv
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcutil/base58"
+
+	"github.com/hyperledger/aries-framework-go/pkg/doc/jose"
+)
+
+const (
+	failComputeMACIndexName         = "failed to compute MAC for index name: %w"
+	failGenerateEDVCompatibleID     = "failed to generate EDV compatible ID: %w"
+	failCreateIndexedAttribute      = "failed to create indexed attribute: %w"
+	failMarshalStructuredDocument   = "failed to marshal structured document into bytes: %w"
+	failEncryptStructuredDocument   = "failed to encrypt structured document into JWE form: %w"
+	failJWESerialize                = "failed to serialize JWE: %w"
+	failMarshalEncryptedDocument    = "failed to marshal encrypted document into bytes: %w"
+	failUnmarshalEncryptedDocument  = "failed to unmarshal encrypted document bytes: %w"
+	failDeserializeJWE              = "failed to deserialize JWE: %w"
+	failDecryptJWE                  = "failed to decrypt JWE: %w"
+	failUnmarshalStructuredDocument = "failed to unmarshal structured document: %w"
+	failToComputeMACIndexValue      = "failed to compute MAC for index value: %w"
+
+	payloadKey  = "payload"
+	keyIndexKey = "indexKey"
+)
+
+var (
+	errPayloadKeyMissing = errors.New(`the structured document content did not contain the ` +
+		`expected "payload" key`)
+	errPayloadNotAssertableAsString = errors.New("unable to assert the payload value as a string")
+)
+
+type marshalFunc func(interface{}) ([]byte, error)
+
+// MACDigester represents a type that can compute MACs.
+type MACDigester interface {
+	ComputeMAC(data []byte, kh interface{}) ([]byte, error)
+}
+
+// MACCrypto is used for computing MACs.
+type MACCrypto struct {
+	kh          interface{}
+	macDigester MACDigester
+}
+
+// ComputeMAC computes a MAC for data using a matching MAC primitive in kh.
+func (m *MACCrypto) ComputeMAC(data string) (string, error) {
+	dataMAC, err := m.macDigester.ComputeMAC([]byte(data), m.kh)
+	return string(dataMAC), err
+}
+
+// NewMACCrypto returns a new instance of a MACCrypto.
+func NewMACCrypto(kh interface{}, macDigester MACDigester) *MACCrypto {
+	return &MACCrypto{
+		kh:          kh,
+		macDigester: macDigester,
+	}
+}
+
+// EncryptedFormatter uses Aries crypto to encrypt and decrypt between
+// Structured Documents and Encrypted Documents.
+type EncryptedFormatter struct {
```

Comment:
The `Formatter` api description mentions the following:
```Formatter represents a type that can encrypt and decrypt between ...```

`EncryptedFormatter` is generic if we assume other `Formatter` implementations will also do encryption/decryption, I would rename this specific type to `JWEFormatter`

Reply:
I would also rectify the description to mention:
```uses JWE to encrypt and decrypt between ...```

Reply:
The description I had on the `Formatter` interface was incorrect... I've updated it.

The `Formatter` represents a type that can convert data from one representation into another. `edv.EncryptedFormatter` is an implementation of this that does plaintext<-->EDV encrypted documents.

Does this clear things up? I think your original suggestion won't apply anymore.

Reply:
Even though locally I updated the comment above the Formatter interface, it looks like I somehow forgot to push it up. I'll fix it in my next PR.